File: 백준/Gold/4803. 트리/트리.ts
const input: number[][] = require("fs").readFileSync(process.platform==="linux"?0:"input.txt").toString().trim().split("\n").map((x:string)=>x.split(" ").map(Number));
let i: number = 0;
let treeNum: number;
let caseNum: number = 0;

while (input[i][0] !== 0) {
    // create graph
    const [n, m] = input[i];
    const graph: Record<number, number[]> = {};
    const visit: number[] = Array(n+1).fill(0);
    for (let j=1; j<n+1; j++) {
        graph[j] = [];
    }
    for (let j=i+1; j<i+m+1; j++) {
        const [u, v] = input[j];
        graph[u].push(v); 
        graph[v].push(u); 
    }
    i += m+1;
    treeNum = 0;
    caseNum += 1;

    // dfs for treeNum
    function bfs(u: number) {
        const stack: number[] = [u];
        let flag = 1;
        visit[u] = 1;
        while (stack.length) {
            const u = stack.pop() as number;
            for (const v of graph[u]) {
                if (visit[v]) {
                    if (visit[v] === visit[u] - 1) continue; // prev node
                    else flag = 0; continue // cycle
                }
                stack.push(v);
                visit[v] = visit[u] + 1;
            }
        }
        return flag;
    }
    for (let u=1; u<n+1; u++) {
        if (visit[u]) continue;
        treeNum += bfs(u)
    }

    // output   
    if (treeNum === 0) console.log(`Case ${caseNum}: No trees.`);
    else if (treeNum == 1) console.log(`Case ${caseNum}: There is one tree.`);
    else console.log(`Case ${caseNum}: A forest of ${treeNum} trees.`);
}

File: 백준/Gold/1068. 트리/트리.py
N = int(input())
arr = [*map(int, input().split())]

G = [set() for _ in range(N)]
P = list(range(N))
exist = [1] * N

for i in range(N):
    if arr[i] == -1:
        continue
    G[arr[i]].add(i)
    P[i] = arr[i]

M = int(input())
stack = [M]
while stack:
    u = stack.pop()
    G[P[u]].discard(u)
    exist[u] = 0
    for v in G[u]:
        stack.append(v)

answer = 0
for i in range(N):
    if exist[i] and len(G[i]) == 0:
        answer += 1

print(answer)